File: util/clock.lua
local Clock = {}

function Clock:new()
    self.startTime = love.timer.getTime()
    self.pauseTime = nil
end

function Clock:pause()
    self.pauseTime = love.timer.getTime()
end

function Clock:resume()
    local elapsed = self:getElapsed()
    self.startTime = love.timer.getTime() - elapsed
    self.pauseTime = nil
end

function Clock:getElapsed()
    if self.pauseTime then
        return self.pauseTime - self.startTime
    else
        return love.timer.getTime() - self.startTime
    end
end

function Clock:restart()
    local elapsed = self:getElapsed()
    self.startTime = love.timer.getTime()
    return elapsed
end

local Clock__mt = {
    __index = Clock
}

return setmetatable(Clock, {
    __call = function(_, ...)
        local o = setmetatable({}, Clock__mt)
        o:new(...)
        return o
    end
})

File: src/util/clock.ts
import { setFullscreen } from "love.window";

export class Clock {
    private startTime: number;
    private pauseTime: number | null;

    constructor() {
        this.startTime = love.timer.getTime();
        this.pauseTime = null;
    }

    pause() {
        this.pauseTime = love.timer.getTime();
    }

    getElapsed() {
        if (this.pauseTime) {
            return this.pauseTime - this.startTime;
        } else {
            return love.timer.getTime() - this.startTime;
        }
    }

    resume() {
        let elapsed = this.getElapsed();
        this.startTime = love.timer.getTime() - elapsed;
        this.pauseTime = null;
    }

    restart() {
        const elapsed = this.getElapsed();
        this.startTime = love.timer.getTime();
        return elapsed;
    }
}
